File: android/app/src/main/java/com/ticketbox/domain/model/BackgroundTask.kt
package com.ticketbox.domain.model

/**
 * ADR-0030 background_tasks domain model.
 *
 * Status values mirror backend ``BackgroundTaskStatus``. We don't model
 * them as an enum yet because new task_types may add states without
 * breaking older clients — Kotlin sealed-class enforcement would force
 * a rebuild for every new state. String compares against the constants
 * below are deliberate.
 */
data class BackgroundTask(
    val publicId: String,
    val taskType: String,
    val status: String,
    val progressCurrent: Int,
    val progressTotal: Int?,
    val progressMessage: String?,
    val errorCode: String?,
    val errorMessage: String?,
    val createdAt: String,
    val startedAt: String?,
    val completedAt: String?,
    val cancellationRequestedAt: String?,
) {
    val isCancellable: Boolean
        get() = status in setOf(BACKGROUND_TASK_QUEUED, BACKGROUND_TASK_RUNNING) &&
            cancellationRequestedAt == null

    val isTerminal: Boolean
        get() = status in setOf(
            BACKGROUND_TASK_COMPLETED,
            BACKGROUND_TASK_FAILED,
            BACKGROUND_TASK_CANCELLED,
        )
}

const val BACKGROUND_TASK_QUEUED = "queued"
const val BACKGROUND_TASK_RUNNING = "running"
const val BACKGROUND_TASK_COMPLETED = "completed"
const val BACKGROUND_TASK_FAILED = "failed"
const val BACKGROUND_TASK_CANCELLED = "cancelled"

/** 任务 ``error_message`` 超过这个长度就当作不可直出的诊断串（正常中文短错误远短于此）。 */
const val BACKGROUND_TASK_ERROR_MAX_DISPLAY_LEN = 60

// 英文异常 / 堆栈特征：后端 error_message 可能是 ``str(exc)`` / "No handler …" /
// orphan 诊断串等英文技术文案，直出违反 §10（用户面不暴露底层异常）。命中任一即泛化。
private val BACKGROUND_TASK_ERROR_STACK_MARKERS = listOf(
    "Exception",
    "Traceback",
    "Error:",
    " at ",
    ".py",
    "java.",
    "org.",
    "com.",
    "app.",
)

/**
 * 8.3: 判定后端任务 ``error_message`` 是否应套泛化文案而非直出。
 *
 * 规则（保守，宁可泛化也不泄露技术串）——满足任一即泛化：
 * - 空白：无内容可展示。
 * - 含换行：多半是堆栈/多行诊断。
 * - 含英文异常/堆栈特征（见 [BACKGROUND_TASK_ERROR_STACK_MARKERS]）。
 * - 超长（> [BACKGROUND_TASK_ERROR_MAX_DISPLAY_LEN]）：正常中文短错误不会这么长。
 * - 不含任何中文：后端可直出的用户面错误都是我们自己写的中文短句；纯英文/符号串视为诊断。
 *
 * 否则（正常中文短消息）原样展示。纯函数，可单测。
 */
fun shouldGeneralizeTaskError(raw: String): Boolean {
    if (raw.isBlank()) return true
    if (raw.contains('\n')) return true
    if (raw.length > BACKGROUND_TASK_ERROR_MAX_DISPLAY_LEN) return true
    if (BACKGROUND_TASK_ERROR_STACK_MARKERS.any { raw.contains(it) }) return true
    return raw.none { it in '一'..'鿿' }
}

File: backend/app/models/background_task.py
"""ADR-0030 long-task execution model.

`background_tasks` is the in-process progress table used by csv-import,
v1-migration, and future long-running user-triggered work. It is a
cross-cutting infrastructure table (not tied to a single business domain),
which is why it lives in its own module rather than next to expense /
identity / budget.

``tenant_id`` is nullable so system-wide tasks (e.g. v1.0 cut-over
migration) can record progress without binding to a single ledger. User-
scoped tasks (csv import) MUST set ``tenant_id`` so the user can only see
their own ledger's task list.
"""

from __future__ import annotations

from datetime import datetime
from uuid import uuid4

from sqlalchemy import (
    CheckConstraint,
    DateTime,
    Index,
    Integer,
    String,
    Text,
)
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base
from app.services.time_service import now_utc


class BackgroundTask(Base):
    """Single row per user-triggered long task; queried via
    ``GET /api/tasks/{public_id}`` and ``GET /api/tasks``.

    State machine (status column):

    ``queued`` -> ``running`` -> ``completed`` | ``failed`` | ``cancelled``

    Orphan recovery: on startup, active rows are force-failed by default
    because the in-process executor that owned them died with the old process.
    Cloud / multi-worker deployments can configure a grace window for fresh
    heartbeating rows; every runner still has to atomically claim ``queued``.
    """

    __tablename__ = "background_tasks"
    __table_args__ = (
        CheckConstraint(
            "status IN ('queued', 'running', 'completed', 'failed', 'cancelled')",
            name="ck_background_tasks_status_valid",
        ),
        CheckConstraint(
            "progress_current >= 0",
            name="ck_background_tasks_progress_current_non_negative",
        ),
        CheckConstraint(
            "progress_total IS NULL OR progress_total >= 0",
            name="ck_background_tasks_progress_total_non_negative",
        ),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    public_id: Mapped[str] = mapped_column(
        String(36), default=lambda: str(uuid4()), nullable=False, unique=True, index=True
    )
    # NULL = system-wide task (e.g. v1.0 migration); set = ledger-scoped task.
    tenant_id: Mapped[str | None] = mapped_column(String(64), nullable=True, index=True)
    task_type: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    initiated_by_account_id: Mapped[int | None] = mapped_column(Integer, nullable=True, index=True)
    initiated_by_device_id: Mapped[int | None] = mapped_column(Integer, nullable=True)

    status: Mapped[str] = mapped_column(
        String(32), default="queued", server_default="queued", nullable=False, index=True
    )
    progress_current: Mapped[int] = mapped_column(
        Integer, default=0, server_default="0", nullable=False
    )
    progress_total: Mapped[int | None] = mapped_column(Integer, nullable=True)
    progress_message: Mapped[str | None] = mapped_column(Text, nullable=True)

    error_code: Mapped[str | None] = mapped_column(String(64), nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    # JSON-encoded; per-task shape (e.g. csv_import returns rows_imported / errors).
    result_summary_json: Mapped[str | None] = mapped_column(Text, nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=now_utc, nullable=False
    )
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    last_progress_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    cancellation_requested_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )


# Composite indexes — match the two main query patterns
Index(
    "ix_background_tasks_account_created",
    BackgroundTask.initiated_by_account_id,
    BackgroundTask.created_at,
)
Index(
    "ix_background_tasks_status_last_progress",
    BackgroundTask.status,
    BackgroundTask.last_progress_at,
)
